controllers/volumereplicationgroup_controller.go

```diff
@@ -1029,72 +1158,280 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 			string(v.instance.Spec.ReplicationState), v.instance.Name, v.instance.Namespace)
 	}
 }
 
-// So, VolumeReplication controller reflects the status of a VolRep resource as primary
-// only after the promotion of the associated storage volume to primary is successful.
-// Hence, checking whether the primary state is reflected in the status is sufficient
-// to verify whether the VolRep resource is in a healthy state not. However, checking
-// conditions of VolRep status may be required for maintaining correct status of VRG
-// when VRG status is implemented.
-//
-// TODO: If checking conditions of VolRep is necessary for VRG status, add that logic
-//       here when VRG status is implemented.
-func (v *VRGInstance) validateVRStatusAsPrimary(volRep *volrep.VolumeReplication) error {
-	if volRep.Status.State != volrep.PrimaryState {
-		v.log.Info("State is primary and status is not primary")
+// VolumeReplication controller sets volRep.Status.State to primary/secondary
+// only after the successful promotion/demotion of the underlying storage volume
+// consumed by the pvc.It is necessary to get the appropriate condition out of
+// different conditions a VolRep status can have. Doing this may be helpful in
+// getting a more clear view of the VolumeReplication resource's current situation.
+func (v *VRGInstance) validateVRStatus(volRep *volrep.VolumeReplication, state ramendrv1alpha1.ReplicationState) error {
+	var (
+		stateString = "unknown"
+		action      = "unknown"
+	)
+
+	switch state {
+	case ramendrv1alpha1.Primary:
+		stateString = "primary"
+		action = "promoted"
+	case ramendrv1alpha1.Secondary:
+		stateString = "secondary"
+		action = "demoted"
+	}
+
+	volRepCondition := findCondition(volRep.Status.Conditions, volrepController.ConditionCompleted)
+	if volRepCondition == nil {
+		v.log.Info("failed to get the Completed condition from VolumeReplication resource", "volRep",
+			volRep.Name)
+
+		msg := "Failed to get the Completed condition from VolumeReplication resource for pvc."
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
+		return nil
+	}
 
-		return fmt.Errorf("primary VolumeReplication resource status not same %s/%s (status: %s)",
-			volRep.Name, volRep.Namespace, volRep.Status.State)
+	if volRepCondition.Status == metav1.ConditionFalse {
+		// As of now, name of the VolumeReplication resource and pvc that it
+		// represents is same. So, using the name from volRep resource. In
+		// future if things change, some changes might have to be necessary.
+		v.log.Info(fmt.Sprintf("VolumeReplication resource is not %s (%s)", action, volRep.Name))
+
+		msg := fmt.Sprintf("VolumeReplication resource for pvc not %s to %s", action, stateString)
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
+		return nil
 	}
 
-	v.log.Info("State in the spec and status match for primary")
+	v.log.Info(fmt.Sprintf("ConditionCompleted is true for state %s (Name: %s)", stateString, volRep.Name))
```

Comment:
It is also possible here that the volRepCondition.Status == metav1.ConditionUnknown and we are not handling it here.

Reply:
ACK

Currently, VolumeReplication controller sets the condition.Status to either metav1.ConditionTrue or metav1.ConditionFalse. So I guess that is why I added a check for ConditionFalse. 

But, I agree with you. Will add a check volRepCondition.Status == metav1.ConditionUnknown and set the reason for the protected pvc condition to PVCErrorUnknown.